localstack/http/resource.py

```diff
@@ -0,0 +1,100 @@
+"""
+This module enables the resource class pattern, where each respective ``on_<http_method>`` method of a class is
+treated like an endpoint for the respective HTTP method. The following shows an example of how the pattern is used::
+
+    class Foo:
+        def on_get(self, request: Request):
+            return {"ok": "GET it"}
+
+        def on_post(self, request: Request):
+            return {"ok": "it was POSTed"}
+
+
+    router = Router(dispatcher=resource_dispatcher())
+    router.add(Resource("/foo", Foo())
+"""
+from typing import Any, Iterable, Optional, Type
+
+from werkzeug.routing import Map, Rule, RuleFactory
+
+from .router import HTTP_METHODS, route
+
+
+def resource(path: str, host: Optional[str] = None, **kwargs):
+    """
+    Class decorator that turns every method that follows the pattern ``on_<http-method>`` into a route,
+    where the allowed method for that route is automatically set to the method specified in the function name. Example
+    when using a Router with the ``handler_dispatcher``::
+
+        @resource("/myresource/<resource_id>")
+        class MyResource:
+            def on_get(request: Request, resource_id: str) -> Response:
+                return Response(f"GET called on {resource_id}")
+
+            def on_post(request: Request, resource_id: str) -> Response:
+                return Response(f"POST called on {resource_id}")
+
+    This class can then be added to a router via ``router.add_route_endpoints(MyResource())``.
+
+    Note that HEAD requests are automatically routed to ``on_get``. There is currently no way of specifying ``on_head``
+    methods because of how werkzeug works. See https://werkzeug.palletsprojects.com/en/2.2.x/routing/
```

Comment:
Why is that? Werkzeug just falls back to GET if there's no HEAD rule matching:
> If GET is present in the list of methods and HEAD is not, HEAD is added automatically.

This seems to be reasonable, considering the spec of HEAD (it's a GET without the body / the headers of GET).
So, imho, adding the HEAD method if it's present would basically translate the semantics of Werkzeug to the resource concept.
This would also avoid the somewhat clumsy "if HEAD then return 200" in the health resource.

Reply:
I was so sure I tested this, but after our call did another reproducer and it's clear that, when you add rules with the same path, that match `HEAD` methods _before_ the rule for `GET` methods, then it works as expected. if you reverse the order of the `add_route_endpoint` calls, then making a HEAD request to the server will print "calling on_get".

In any case, this just means the resource rules need to be added in the correct order. will fix!

```python
from werkzeug import run_simple, Request, Response

from localstack.http import Router, route


@route("/foo", methods=["HEAD"])
def on_head(request, args):
    print("calling on_head")
    return Response("HEAD")


@route("/foo", methods=["GET"])
def on_get(request, args):
    print("calling on_get")
    return Response("GET")


def main():
    router = Router()

    router.add_route_endpoint(on_head)
    router.add_route_endpoint(on_get)

    run_simple("localhost", 5000, Request.application(router.dispatch))


if __name__ == "__main__":
    main()
```